Widget endpoint docs (#144)
resource-watch/doc-api: source/includes/api/_widget.md

```

> Example response:

```json
{
    "data": [
        {
            "id": "73f00267-fe34-42aa-a611-13b102f38d75",
            "type": "widget",
            "attributes": {
                "name": "Precipitation Change in Puget Sound",
                "dataset": "06c44f9a-aae7-401e-874c-de13b7764959",
                "slug": "precipitation-change",
                "userId": "58333dcfd9f39b189ca44c75",
                "description": "NOAA nCLIMDIV Precipitation: Historical Precipitation",
                "source": null,
                "sourceUrl": null,
                "authors": null,
                "application": [
                    "prep"
                ],
                "verified": false,
                "default": true,
                "protected": false,
                "defaultEditableWidget": false,
                "published": true,
                "freeze": false,
                "env": "production",
                "queryUrl": "query/06c44f9a-aae7-401e-874c-de13b7764959?sql=select annual as x, year as y from index_06c44f9aaae7401e874cde13b7764959%20order%20by%20year%20asc",
                "widgetConfig": {...}
            }
        },
        {...}
    ]
}
```

> This endpoint allows authenticated users to load all widgets belonging to multiple datasets in a single request. 

Reviewer: Why do I have to be authenticated for this one, but not the above?

Author: This text is a copy-paste from the layers docs. For layers, you do need to be authenticated to use this endpoint. For widgets, you can actually use it without being authenticated. "why" will forever remain a mystery, but I suspect these endpoints are meant to be internal, as they don't paginate responses - AKA you can easily timeout or get a multiple MB json reply using them.

Author: I'll update the text so it doesnt say users need to be auth'd

> Return all widgets associated with two datasets, that are associated with either `rw` or `prep` applications 

```shell
curl -X POST "https://api.resourcewatch.org/widget/find-by-ids" \
-H "Authorization: Bearer <your-token>" \
-H 'Content-Type: application/json' \
-d '{
	"ids": ["<dataset 1 id>", "<dataset 2 id>"],

>     "app" "rw,prep"

Reviewer: ```suggestion
    "app": "rw,prep"
```

}'
```

> Return all widgets associated with two datasets, that are associated with both `rw` and `prep` applications simultaneously 

```shell
curl -X POST "https://api.resourcewatch.org/widget/find-by-ids" \
-H "Authorization: Bearer <your-token>" \
-H 'Content-Type: application/json' \
-d '{
	"ids": ["<dataset 1 id>", "<dataset 2 id>"],

>     "app" "rw@prep"

Reviewer: ```suggestion
    "app": "rw@prep"
```

>     "app" "rw@prep"

Reviewer: This filter by application syntax looks very similar to the layer endpoint...is it worth factoring out into a more general section, or is it just in these two places?

```

Field        |         Description          |   Type |   Default
------------ | :--------------------------: | -----: | ----------:
page[size]   | The number elements per page. Values above 100 are not officially supported. | Number | 10
page[number] |       The page number        | Number | 1

### Filters

> Return the widgets filtered by those whose name contains emissions

```shell
curl -X GET "https://api.resourcewatch.org/v1/widget?name=emissions"
```

> Return the widgets filtered by dataset

```shell
curl -X GET "https://api.resourcewatch.org/v1/widget?dataset=11de2bc1-368b-42ed-a207-aaff8ece752b"
curl -X GET "https://api.resourcewatch.org/v1/dataset/11de2bc1-368b-42ed-a207-aaff8ece752b/widget"
```

> Filter widgets by default value

```shell
curl -X GET "https://api.resourcewatch.org/v1/widget?default=false"
```

> Filter widgets by environment

```shell
curl -X GET "https://api.resourcewatch.org/v1/widget?env=staging"
```

> Return the widgets filtered by those whose applications contain both `rw` and `prep` applications simultaneously

```shell
curl -X GET "https://api.resourcewatch.org/v1/widget?app=rw@prep"
```

> The widget list endpoints provide a wide range of filters that you can use to tailor your widget listing. The great majority of the widget fields you'll find on the [widget reference](#widget-reference) section can be used as filters when listing widgets, with the following exceptions:

Reviewer: Is it actually all of them except these three? If that's the case I would just say "all" rather than "the great majority".

- `id`
- `userName`
- `userRole`: filtering by the role of the owning user can be done using the `user.role` query argument.

Additionally, you can use the following filters:

- `collection`: filters by a [collection](#collections) id. Requires being authenticated.

> - `favourite`: if defined, only the widgets set by the user as [favorites](#favorites) will be returned. Requires being authenticated.

Reviewer 2: Is this a boolean filter? (if so, that should be in the description of the filter)

Author: I'll clarify the language, thanks

Multiple parameters can be combined into a complex `and` logic filter. Depending on the type of the field you're filtering by, the following behavior is available:

- String type fields support both regular expressions and partial matches.
- Array type fields support `,` as `OR` and `@` as `AND` separators for multiple values.

> - Object type fields will match the whole object, not just its parts.

Reviewer: And list types too, I assume?

Author: sorry, I don't understand what you mean by "list" here. Are you perhaps referring to arrays? those are covered on the previous line.

Reviewer: Yes sorry just didn't read carefully

### Sorting

> The API currently supports sorting by means of the `sort` parameter. Sorting can be done using any field from the widget, as well as `user.name` and `user.role` (sorting by user data is restricted to `ADMIN` users).

Reviewer: Same question as for filtering - can we separate this out into its own more general sorting section?

Author: seems like this sorting logic is used on datasets, layers, widgets and metadata. I suspect it's not used anywhere else - I checked vocabulary, area, geostore and subscriptions, and none of them have it. 

I honestly have no good reason to pick one over the other - centralizing is better if things are consistent (I need to double-check this) but it means more work if things diverge. Engineer laziness has a slight preference towards keeping things as is, but I'll leave it up to you to decide

Reviewer: Yeah I guess it depends on consistency - is the sorting/filtering logic implemented separately for all of these resources, or is it a common layer on top?

If it's consistent, I would add a "Sorting & Filtering" section at the top and then have a stub in each of the resources that supports it that says please see the sorting and filtering section above.

Author: No, each MS has its own specific code for sorting, it just happens to be copy-pasted from one to another - to the point where there's a bug in that code, that now only exists in metadata because I fixed it for dataset widget and layer as I wrote these docs.

I take it then that we should leave it as is in the docs?

Reviewer: Yeah let's leave it...potentially could consider fixing these bugs and implementing generic sorting and filtering middleware for the next contract

> Loads related vocabularies. If none are found, an empty array is returned.

Reviewer 2: It might make sense to add here a link to the vocabulary documentation :)

            "published": true,
            "freeze": false,
            "env": "production",
            "queryUrl": "/v1/query?sql=Select * from data",
            "widgetConfig": {
                "data": {
                    "url": "/v1/query?sql=Select * from data"
                }
            },
            "template": false,
            "layerId": null,
            "createdAt": "2017-02-08T15:30:34.505Z",
            "updatedAt": "2017-02-08T15:30:34.505Z"
        }
    }
}
```

When getting a single widget, you can optionally provide a `queryUrl` query parameter. If provided, this parameter will overwrite the following response values:

- `data.attributes.queryUrl`
- `data.attributes.widgetConfig.data.url`
- `data.attributes.widgetConfig.data[0].url`
 
This overwrite will only happen if a previous value already existed for the respective field. Also, keep in mind that these changes will NOT be persisted to the database, and will only affect the current request's response. 
 

> ### Customize the query url of a widget response

Reviewer: Huh..what's the purpose of this (and the above overwrite functionality)?

Author: @aagm tells me that this was requested to make widgets more customizeable - instead of having different widgets for variations of the same visualization, this allows tweaking the same widget without duplicating it. As to why this logic is in the API as opposed to the client apps using it - the best reason we could come up with was to avoid having to duplicate this logic on potentially multiple clients.

            "layerId": null,
            "createdAt": "2017-02-08T15:30:34.505Z",
            "updatedAt": "2017-02-08T15:30:34.505Z"
        }
    }
}
```

When getting a single widget, you can optionally provide additional custom query parameter. These parameters will be appended to the following response values:

- `data.attributes.queryUrl`
- `data.attributes.widgetConfig.data.url`
- `data.attributes.widgetConfig.data[0].url`

The widget service assumes these values are URLs, and will append your custom query parameters as such:

- If the value already has a `?` character present, your custom query parameters will be appended with `&` preceding them. 
- If no `?` is present, the first custom query parameter will be appended with a `?`, and the following with a `&` before them.

This overwrite will only happen if a previous value already existed for the respective field. Also, keep in mind that these changes will NOT be persisted to the database, and will only affect the current request's response. You can combine these custom query parameters with an [overwritten query url](#overwrite-the-query-url-of-a-widget-response), as seen on the included example.

> ### Include entities associated with the widget

Reviewer: This section is duplicated, no?

Author: we have one above for the "get multiple widgets" endpoint. this one is under the "get a single widget by id" endpoint, and it just points to the other one. The idea is to keep the docs DRY, but at the same time cover users who go straight for the "get single widget by id" without reading what came before.

Reviewer: Ah okay. I would maybe just say something like, "As with getting all widgets, you can include related entities for a single widget," and then link. Not a big deal though.

Author: I think that's the same that's already on line 735, no?

-H "Authorization: Bearer <your-token>" \
-H "Content-Type: application/json"
```

> Response

```json
{
    "data": {
        "id": "7946bca1-6c2a-4329-a127-558ef9551eba",
        "type": "widget",
        "attributes": {
            "name": "name for the cloned widget",
            "description": "description of the cloned widget",
            "dataset": "7fa6ec77-5ab0-43f4-9a4c-a3d19bed1e90",
            "slug": "Example-Widget_2",
            "userId": "5dbadb0adf2dc74d2ad05dfb",
            "application": [
                "gfw"
            ],
            "verified": false,
            "default": false,
            "protected": false,
            "defaultEditableWidget": false,
            "published": true,
            "freeze": false,
            "env": "production",
            "template": false,
            "createdAt": "2020-06-11T14:13:19.677Z",
            "updatedAt": "2020-06-11T14:13:19.677Z"
        }
    }
}
```

Use this endpoint if you'd like to delete a widget from the RW API. As a widget object may not store the actual data being displayed, this may only delete the widget settings, but the actual data may continue to be available at its source.

> Besides deleting the widget itself, this endpoint also deletes graph vocabularies and metadata related to the widget itself. These delete operations are issued after the widget itself is deleted. The process is not atomic, and the output of the API request is based solely on the result of the deletion of the widget itself. For example, is the metadata service is temporarily unavailable when you issue your delete widget request, the widget itself will be deleted, but the associated metadata will continue to exist. The response will not reflect the failure to delete metadata in any way.

Reviewer: ```suggestion
Besides deleting the widget itself, this endpoint also deletes graph vocabularies and metadata related to the widget itself. These delete operations are issued after the widget itself is deleted. The process is not atomic, and the response is based solely on the result of the deletion of the widget itself, not related resources. For example, is the metadata service is temporarily unavailable when you issue your delete widget request, the widget itself will be deleted, but the associated metadata will continue to exist. The response will not reflect the failure to delete metadata in any way.
```

userId                  | String         | Yes (autopopulated)  |                            | Id of the user who created the widget. Set automatically on creation. Cannot be modified by users.
description             | String         | No                   |                            | User defined description of the widget.
source                  | String         | No                   |                            | Description of the source of the widget's data, as it's meant to be displayed to the end user.
sourceUrl               | String         | No                   |                            | URL of the source of the data, as it's meant to be displayed to the end user.
authors                 | String         | No                   |                            | Author or authors of the data displayed on the widget, as it's meant to be displayed to the end user.
queryUrl                | String         | No                   |                            | URL of the RW API query or external URL containing the data displayed on the widget

> thumbnailUrl            | String         | No                   |                            | URL of a example thumbnail of the rendered widget.

Reviewer 2: Can the `thumbnailUrl` property be set by end-users?

Author: technically it can be updated by the user, but the update process triggers a thumbnail autogeneration, so it won't hold the user defined value for long
